narwhal/primary/src/core.rs

```diff
@@ -185,327 +173,260 @@ impl Core {
+                    .as_secs_f64(),
+            );
 
-        // Check if we can vote for this header.
-        // Send the vote when:
-        // 1. when there is no existing vote for this publicKey & round
-        //    (for this publicKey the last voted round in the votes store < header.round)
-        // 2. when there is a vote for this publicKey & round,
-        //    (for this publicKey the last voted round in the votes store = header.round)
-        //    and the hash also corresponding to the vote we already sent matches the hash of
-        //    the vote we create for this header we received
-        // Taking the inverse of these two, the only time we don't want to vote is when:
-        // there is a digest for the publicKey & round, and it does not match the digest of the
-        // vote we create for this header.
-        // Also when the header round is less than the latest round we have already voted for,
-        // then it is useless to vote, so we don't.
-
-        let result = self
-            .vote_digest_store
-            .read(header.author.clone())
-            .await
-            .map_err(StoreError)?;
-
-        if let Some(round_digest_pair) = result {
-            if header.round < round_digest_pair.round {
-                return Ok(());
-            }
-            if header.round == round_digest_pair.round {
-                // check the hash first
-                let temp_vote = Vote::new(header, &self.name, &mut self.signature_service).await;
-                if temp_vote.digest() != round_digest_pair.vote_digest {
-                    // we already sent a vote for a different header to the authority for this round
-                    // don't equivocate by sending a different vote for the same round
-                    warn!(
-                        "Authority {} submitted duplicate header for votes at round {}",
-                        header.author, header.round
-                    );
-                    self.metrics
-                        .votes_dropped_equivocation_protection
-                        .with_label_values(&[&header.epoch.to_string(), header_source])
-                        .inc();
-                    return Ok(());
-                }
-            }
-        }
-        self.send_vote(header).await
-    }
-
-    #[instrument(level = "debug", skip_all)]
-    async fn send_vote(&mut self, header: &Header) -> DagResult<()> {
-        // Make a vote and send it to the header's creator.
-        let vote = Vote::new(header, &self.name, &mut self.signature_service).await;
-        debug!(
-            "Created vote {vote:?} for {header} at round {}",
-            header.round
+        // Broadcast the certificate.
+        let network_keys = self
+            .committee
+            .others_primaries(&self.name)
+            .into_iter()
+            .map(|(_, _, network_key)| network_key)
+            .collect();
+        self.network.unreliable_broadcast(
```

Comment:
Yes, while we have not advanced in terms of narwhal round, we should try and re-try forming a certificate for the current round header.

Reply:
There are two separate topics here I think. In terms of continually trying to form a certificate for current round, that is driven by proposer right? I believe the proposer should keep sending the Core headers for the current round until a certificate is successfully formed (and then sent back in the below call to `process_certificate_internal`). Similarly, the Core will continually try to get votes to form a certificate for each proposed header until either (1) a cert is formed, (2) a cert is deemed impossible to form due to >f unrecoverable errors, or (3) a new header is received from proposer.

As for broadcasting the successfully formed cert to everyone else, can you say more about what would result in the loss of liveness? Best effort seemed sufficient here because any node that doesn't happen to receive it in the initial broadcast will still hear about it when they receive a vote request for a Header that names the new certificate as parent, and at that time they will ask us for it. Even if no other node lists our cert as parent (unlikely but possible), then we still will do it and send others those vote requests, I think?

My philosphy here was to make everything as request-driven as possible. It's fine to broadcast things on a best-effort basis, but reliable broadcast should be used very sparingly because we don't even know if the nodes we are broadcasting to want or would find the thing we are sending them useful. For example maybe they already got our certificate from another node, if we are still trying to send it to them minutes later. But if they are requesting it from us explicitly (eg via a RequestVote missing parents indication) then we for sure know that they want it and it'd be useful to them.

Reply:
It's worth noting here too that `unreliable_broadcast` should still work successfully almost all the time during normal conditions, and during abnormal conditions (node is behind, network issues, etc.) is exactly when you probably do NOT want to be clogging things up with Certificate broadcast retries, because that will then compete for resources with more important catch-up processes.

Reply:
> I believe the proposer should keep sending the Core headers for the current round until a certificate is successfully formed 
The proposer doesn't do this currently, so we shouldn't make a change while relying that is does this now. There are tradeoffs to be considered with that change as well, so for now let's not expand the scope of this one PR to make that change. 

> Similarly, the Core will continually try to get votes to form a certificate for each proposed header until either (1) a cert is formed, (2) a cert is deemed impossible to form due to >f unrecoverable errors, or (3) a new header is received from proposer.

Can you point me to the place in this PR where (2) causes us to stop trying? What kind of errors are these?

> As for broadcasting the successfully formed cert to everyone else, can you say more about what would result in the loss of liveness? Best effort seemed sufficient here because any node that doesn't happen to receive it in the initial broadcast will still hear about it when they receive a vote request for a Header that names the new certificate as parent, and at that time they will ask us for it. Even if no other node lists our cert as parent (unlikely but possible), then we still will do it and send others those vote requests, I think?

The certs are an invariant for liveness, in particular they lead to the proposer being able to tell the core to form the next header. If we were to never hear about some certs due to the unreliable broadcast, we would have cases where the proposer didn't get enough certs to trigger the header creation, and it wouldn't know that that happened.

> My philosphy here was to make everything as request-driven as possible. It's fine to broadcast things on a best-effort basis, but reliable broadcast should be used very sparingly because we don't even know if the nodes we are broadcasting to want or would find the thing we are sending them useful. For example maybe they already got our certificate from another node, if we are still trying to send it to them minutes later. But if they are requesting it from us explicitly (eg via a RequestVote missing parents indication) then we for sure know that they want it and it'd be useful to them.

I don't think cert delivery makes sense as request driven because that will lead to either constant polling or a much slower execution speed.

Reply:
I think that unreliable sending is ok here, only if the proposer part of the core within a round keeps re-sending its header and then its certificate to others, until the round advances. So instead of making each message reliable we just keep repeating the process needed to make progress at a more macro level, until it succeeds. Not sure what approach is simpler.

But what is sure is, that if we do not ensure that eventually all receive Header, Votes and Certs from other honest nodes, then we will lose liveness.

Reply:
I updated this to continually retry certificate broadcast until either it succeeds or the narwhal round advances. Ty for the feedback!